### DISCUSSION TITLE
Component library

### DISCUSSION THREAD

**Discussion by asherccohen** [tayv/Project-Bubblegum#41] [Ideas]
The repo is firstly a component library to be consumed by other apps.

It requires bundling configuration and ability to be published to npm.

Meanwhile the structure can be readapted.

Suggestions:

- atomic design
- storybook
- theming
- JEST tests
- code styling conventions (prettier, linter, commit hooks)

**Comment by tayv** [comment]
I like this idea and agree it's the best route for the stated goal. However, I do wonder if I'd be getting ahead of myself trying to put together a full blown design system/library. 

Some concerns: 

- Storybook
    - Wrestling with configuration, esp. with RHF controlled components. For example https://github.com/react-hook-form/react-hook-form/discussions/7444
    - Opportunity cost of setup time and learning syntax
    - Useful but perhaps not this early on

- JEST tests
  - Opportunity cost of time setting up. I agree these are important but given I'll like be tearing down and refactoring my app and UI components a few times it seems like this might be better further down the road
  
- NPM bundling
  - Same concern as above re: opportunity cost of time if I'm likely to be refactoring the components 
  - The need to follow semver also seems like it could slow down refactors
  
Perhaps best to put a pin in these more systematic approaches until the components and their use cases are more fleshed out / battle tested?

**Reply by asherccohen** [reply]
Storybook we can skip, but helps a lot to visually detect issues.

Jest is fundamental, you will wish to have started with tests soon after a few refactorings.

The nice part of the above is that Storybook can do jest testing and you've got two points with one config.

Plus you can build/test up to full pages in isolation (something most devs overlook, thinking it's only for components)

I agree Npm bundling is not a prio until the repo is a next app and custom libs are consumed within it. Just keep it on the roadmap.
